### ISSUE TITLE
🐞[Enter Bug Description] Cannot register OneOffTask iOS

### ISSUE THREAD

**Issue by saad197** [fluttercommunity/flutter_workmanager#357]
- [x] I have read the [README](https://github.com/vrtdev/flutter_workmanager/blob/master/README.md)
- [x] I have done the setup for [Android](https://github.com/vrtdev/flutter_workmanager/blob/master/ANDROID_SETUP.md)
- [x] I have done the setup for [iOS](https://github.com/vrtdev/flutter_workmanager/blob/master/IOS_SETUP.md)
- [x] I have ran the sample app and it does not work there

**Version**

| Technology           | Version       |
| -------------------  | ------------- |
| Workmanager version  |        0.5.0-dev.8       |
| Xcode version        |       13.2.1        |
| Swift version        |       5.5.2-dev        |
| iOS deployment target|        13.0       |

**Describe the error**
I am not able to schedule OneOff task on iOS. I can only do it if I through debug, "Simulate background fetch" which gets me

 `2022-02-01 01:09:06.770562-0700 Runner[15798:4641652] flutter: notifications set ios
2022-02-01 01:09:06.853857-0700 Runner[15798:4640548] [workmanager.BackgroundWorker] performBackgroundRequest(_:) -> performBackgroundRequest.newData (finished in 0.48 seconds)`

I know everything is set up correctly because simulation would not work otherwise...

The same issue on Workmanager 0.4.1 except, here is the interesting part, 0.4.2 would give me this error when I would register my task 

`PlatformException(unhandledMethod("registerOneOffTask") error, Unhandled method registerOneOffTask, null, null) #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:607:7) #1      MethodChannel._invokeMethod`

 

**Output of `flutter doctor -v`**

[✓] Flutter (Channel stable, 2.8.1, on macOS 12.1 21C52 darwin-x64, locale en-CA)
    • Flutter version 2.8.1 at /users/saad/Documents/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (7 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Users/saad/Library/Android/sdk
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7281165)

[✓] VS Code (version 1.63.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

**Comment by Iam-that-Chic** [comment]
following, i have the same error too.

**Comment by cormalenv** [comment]
I'm a bit confused by this. Is it actually possible to register one of tasks on iOS? I was under the impression this is an Android only function as described in the [ReadMe](https://github.com/fluttercommunity/flutter_workmanager/blob/main/README.md) (although I note the example triggers `registerOneOffTask` via a platform enable iOS button?!?).

The [API](https://pub.dev/documentation/workmanager/latest/workmanager/Workmanager-class.html) states that once Workmanager is initialised you can then `registerOneOffTask` or `registerPeriodTask` **_for Android_** whereas " **_iOS periodic task is automatically scheduled_**". The API therefore appears consistent with the tasking specifications  for iOS. For example iOS controls timing of the 3 available tasking options; `AppRefresh`, `Processing` and `URLSession`_. As I understand Workmanager utilises _`Processing`_ for which triggering is ultimately out of the developers hands.

I am still not sure why the example includes a `registerOneOffTask` then. Sorry if I have confused this further.....

**Comment by saad197** [comment]
Yes so currently it is impossible to schedule anything on iOS, so workmanager basically does not work.

**Comment by orestesgaolin** [comment]
Interestingly there's entry in flutter log indicating that the given task was _received_, for instance for id `1111` the log says:

```
flutter: Received 1111
```

**Comment by cormalenv** [comment]
@saad197 IMHO sort of....

From what I can tell forget timed scheduling/one of tasks in iOS. Initialising the background service and handing off to the OS is all that can be achieved. This isn't a `workmanager` specific issue this is all about iOS. All `workmanager` can do is provide a method to support either `BGAppRefreshTask` or `BGProcessingTask`. Once this is done it is over to the OS to schedule and your app must be developed to cope with as much.

If you put aside any notion about achieving set intervals , `workmanager` does appear to function correctly under the limitations of the iOS (`oneOffTasks` aside). Initialising the service and simulating a background fetch works fine. The plugin uses `BGProcessingTask` (as far as I can tell), so getting the task to run outside of simulation is another story. To find out more about how your `BGProcessingTask` is managed  and scheduled in the real world skip through to 15:00 [in this video from apple](https://developer.apple.com/videos/play/wwdc2020/10063/).

The documentation probably doesn't spell this out enough and there appear some contradictions that need to be cleared up.

It's clearly a confusing topic and seeing semi-regular periodic tasks work fine on Android doesn't help. The `background_fetch` plugin tries to clarify this a bit better. Incidentally it promotes the use of `BGAppRefreshTask` and suggests that `BGProcessingTask` is bit of hit and miss, almost recommending not to use it but each have their place. You just need to determine what suits your purpose and develop to cope with the limitations on the OS.

I say all this IMHO. I am very happy to be corrected.

**Comment by saad197** [comment]
@cormalenv so my question is how do I make the plugin work at all with iOS? Because as stated even when initialized without any errors it does not actually register or execute the task, ever. I tested this on my physical iPhone.

**Comment by cormalenv** [comment]
> @cormalenv so my question is how do I make the plugin work at all with iOS? Because as stated even when initialized without any errors it does not actually register or execute the task, ever. I tested this on my physical iPhone.

@saad197 How long did you let it run on the physical device? I have found it can take a day or two for iOS to analyse use patterns before it settles on appropriate timings and kicks off the first execution.

**Comment by rackberg** [comment]
Hello there,

we faced the same issue you've described and discussed above.

After digging into the code and also reading through the ios code of the provided example app we noticed that the documentation in the README file is missing a very important step for the ios setup.

Insert this in your AppDelegate of your flutter app:
```
WorkmanagerPlugin.setPluginRegistrantCallback { registry in
    // Registry in this case is the FlutterEngine that is created in Workmanager's
    // performFetchWithCompletionHandler or BGAppRefreshTask.
    // This will make other plugins available during a background operation.
    GeneratedPluginRegistrant.register(with: registry)
}
```
For reference: This code is from the AppDelegate of the example app.

For every task identifier registered under the key `BGTaskSchedulerPermittedIdentifiers` in your app Info.plist file 
add the following line of code to the AppDelegate:

For demonstration purpose we assume that the task e.g. "my.task" is set in the Info.plist file:
```
WorkmanagerPlugin.registerTask(withIdentifier: "my.task")
```

After doing that the callbackDispatcher you used in the `Workmanager().initalize()` method should be called
when you execute the background fetch simulation of xCode under "Debug" > "Simulate Background Fetch"

This should immediately execute the background task while debugging with xCode.

**Comment by ened** [comment]
Please add a PR for missing docs

**Comment by EA-YOUHOU** [comment]
Not working  : ```
 type 'WorkmanagerPlugin' has no member 'registerTask'
 ```

**Comment by leonus96** [comment]
> Didn't know where to put
> 
> ```
> SystemCapabilities = {
> 	com.apple.BackgroundModes = {
> 		enabled = 1;
> 	};
> };
> ```
> 
> project.pbxproj had more than 4800 lines.

It was not necessary for me

**Comment by KirioXX** [comment]
Has anyone a solution for this issue?
I pretty much copied the example app but I still get this error:
```sh
PlatformException (PlatformException(bgTaskSchedulingFailed(Error Domain=BGTaskSchedulerErrorDomain Code=1 "(null)") error, Scheduling the task using BGTaskScheduler has failed.

This may be due to too many tasks being scheduled but not run.

See the error for details: Error Domain=BGTaskSchedulerErrorDomain Code=1 "(null)"., null, null))
```

**Comment by KirioXX** [comment]
Hi, I see similar behaviour on the example app.
`registerOneOffTask` is doing nothing,
but when I trigger "Simulate background fetch" the app crashes with these logs:
```sh
2022-08-23 09:58:10.185987+0100 Runner[2719:70633] Metal API Validation Enabled
2022-08-23 09:58:10.483265+0100 Runner[2719:70872] flutter: The Dart VM service is listening on http://127.0.0.1:58076/C7PhGHlkerc=/
2022-08-23 09:59:05.105665+0100 Runner[2719:71113] flutter: The iOS background fetch was triggered
2022-08-23 09:59:05.111667+0100 Runner[2719:71113] flutter: You can access other plugins in the background, for example Directory.getTemporaryDirectory(): /var/mobile/Containers/Data/Application/AA9E7604-3B91-4B80-A7B9-A49CCBE39E28/Library/Caches
2022-08-23 09:59:05.280439+0100 Runner[2719:70633] [workmanager.BackgroundWorker] performBackgroundRequest(_:) -> performBackgroundRequest.newData (finished in 0.47 seconds)
```

My system:
* MacBook Pro (intel version)
* MacOS Monterey (12.5.1)

My test device:
* Device: iPad (7th generation)
* iOS version: 15.6.1

`flutter doctor -v` output:
```sh
[✓] Flutter (Channel stable, 3.0.5, on macOS 12.5.1 21G83 darwin-x64, locale en-GB)
    • Flutter version 3.0.5 at /Users/___/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision f1875d570e (6 weeks ago), 2022-07-13 11:24:16 -0700
    • Engine revision e85ea0e79c
    • Dart version 2.17.6
    • DevTools version 2.12.2

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/___/Library/Android/sdk
    • Platform android-32, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.4.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7590822)

[✓] VS Code (version 1.70.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.46.0

[✓] Connected device (1 available)
    • My iPad (mobile) • 2783e9a7ba84ac8449beef783bd2d872ea5272f6 • ios            • iOS 15.6.1 19G82

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```
Is there something I could miss in the setup or should the example work out of the box?

**Comment by teenrage-dev** [comment]
Do you have any updates?

**Comment by ened** [comment]
This predates several iOS reworks. One-off/periodic iOS support was overhauled in the 0.5.1+ releases (BGTaskScheduler support, #511) and again in the current federated architecture: #687 passes taskName and honors initialDelay, #692 auto-registers launch handlers so Dart-scheduled tasks survive app relaunch, and the Quick Start documents the required registrant wiring (#694). The current package is `workmanager_apple` (0.9.3+1). Please retest on current versions; reopening with a repro is welcome if tasks still can't be registered.
